File: bench/iter_chain/iter_chain.java
// Java peer of bench/iter_chain.vader. Uses the Stream API — lazy by
// default, dispatches one element at a time through the pipeline,
// mirroring Vader's MapIterator + FilterIterator chain.

static final int N = 1_000_000;

void main() {
    long total = IntStream.range(0, N)
        .filter(x -> x % 2 == 0)
        .mapToLong(x -> (long) x * (long) x)
        .sum();
    IO.println("iter_chain N=%d total=%d".formatted(N, total));
}

File: bench/iter_chain/iter_chain.ts
// TypeScript peer of bench/iter_chain.vader. Uses generator functions for
// a lazy chain that mirrors Vader's MapIterator + FilterIterator + Range —
// per-item allocation of the yielded {value, done} pair, three-level
// dispatch through `for…of`.

const N = 1_000_000;

function* range(n: number): Generator<number> {
  for (let i = 0; i < n; i++) yield i;
}

function* filterGen<T>(it: Iterable<T>, pred: (x: T) => boolean): Generator<T> {
  for (const x of it) if (pred(x)) yield x;
}

function* mapGen<T, U>(it: Iterable<T>, f: (x: T) => U): Generator<U> {
  for (const x of it) yield f(x);
}

const isEven = (x: number): boolean => x % 2 === 0;
const squareI64 = (x: number): bigint => BigInt(x) * BigInt(x);

let total = 0n;
for (const x of mapGen(filterGen(range(N), isEven), squareI64)) {
  total += x;
}
console.log(`iter_chain N=${N} total=${total}`);
